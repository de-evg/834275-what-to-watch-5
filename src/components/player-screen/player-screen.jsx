import React, {useCallback, useRef, useState} from "react";
import {connect} from "react-redux";
import {Link} from "react-router-dom";
import playerScreenProps from "./player-screen.props";

const MAX_DURATION_PERCENT = 100;

const PlayerScreen = (props) => {
  const initialState = {
    duration: 0,
    currentPercent: 0,
    playedTime: 0,
    isPlaying: false
  };
  const [playerProperties, setPlayerProperties] = useState(initialState);
  const {playedTime, currentPercent, isPlaying} = playerProperties;
  const videoRef = useRef();

  const togglePlayPause = useCallback(() => {
    const video = videoRef.current;
    return video.paused ? video.play() : video.pause();
  }, [videoRef]);

  const changeToFullScreen = useCallback(() => {
    videoRef.current.requestFullscreen();
  }, [videoRef]);

  const handelBtnPlayClick = useCallback(() => {
    togglePlayPause(videoRef.current);
    setPlayerProperties(Object.assign(
        {},
        playerProperties,
        {isPlaying: !isPlaying}
    ), [setPlayerProperties, playerProperties, isPlaying]);
  });

  const handleVideoClick = useCallback((evt) => {
    if (evt.target.tagName === `VIDEO`) {
      togglePlayPause();
      setPlayerProperties(Object.assign(
          {},
          playerProperties,
          {isPlaying: !isPlaying}
      ), [setPlayerProperties, playerProperties, isPlaying]);
    }
  });

  const handleDurationChange = useCallback((evt) => {
    const {currentTime} = evt.target;
    const percent = runTime
      ? currentTime * MAX_DURATION_PERCENT / runTime
      : 0;
    setPlayerProperties(Object.assign(
        {},
        playerProperties,
        {
          currentPercent: percent,
          playedTime: currentTime
        }
    ), [setPlayerProperties, playerProperties, runTime]);
  });

  const handleMoiveCanPlay = useCallback(() => {
    setPlayerProperties(Object.assign(
        {},
        playerProperties,
        {playedTime: runTime}
    ), [setPlayerProperties, playerProperties, runTime]);
  });

  const handleFullScreenChange = useCallback(() => {
    changeToFullScreen();
  });

  const {movies, match: {params: {id}}} = props;
  const currentMovie = movies.find((movie) => movie.id === +id);
  const {videoURL, previewURL, title, runTime} = currentMovie;
  const hours = Math.floor(runTime / 3600) - Math.floor(playedTime / 3600);
  const minutes = Math.floor(runTime % 3600 / 60) - Math.floor(playedTime % 3600 / 60);
  const seconds = Math.floor(runTime % 3600 % 60) - Math.floor(playedTime % 3600 % 60);

  return (
    <div className="player">
      <video
        ref={videoRef}
        className="player__video"
        poster={previewURL}
        src={videoURL}
        onClick={handleVideoClick}
        onTimeUpdate={handleDurationChange}
        onCanPlay={handleMoiveCanPlay}
        autoPlay={true}
      />

      <Link to={`/`} type="button" className="player__exit" style={{textDecoration: `none`}}>Exit</Link>

      <div className="player__controls">
        <div className="player__controls-row">
          <div className="player__time">
            <progress className="player__progress" value={currentPercent} max="100"></progress>
            <div className="player__toggler" style={{left: `${currentPercent}%`}}>Toggler</div>
          </div>
          <div className="player__time-value">{`${hours}:${minutes}:${seconds}`}</div>
        </div>

        <div className="player__controls-row">
          <button onClick={handelBtnPlayClick} type="button" className="player__play">
            {
              isPlaying
                ? (
                  <>
                    <svg viewBox="0 0 19 19" width="19" height="19">
                      <use xlinkHref="#play-s" />
                    </svg>
                    <span>Play</span>
                  </>
                )
                : (
                  <>
                    <svg viewBox="0 0 14 21" width="14" height="21">
                      <use xlinkHref="#pause"></use>
                    </svg>
                    <span>Pause</span>
                  </>
                )
            }
          </button>
          <div className="player__name">{title}</div>

          <button onClick={handleFullScreenChange} type="button" className="player__full-screen">
            <svg viewBox="0 0 27 27" width="27" height="27">
              <use xlinkHref="#full-screen" />
            </svg>
            <span>Full screen</span>
          </button>
        </div>
      </div>
    </div>
  );
};

PlayerScreen.propTypes = playerScreenProps;

const mapStateToProps = (state) => ({
  movies: state.DATA.movies
});

export {PlayerScreen};
export default connect(mapStateToProps)(PlayerScreen);
