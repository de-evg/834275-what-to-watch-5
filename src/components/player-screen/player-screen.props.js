import PropTypes from "prop-types";
import {movieProps} from "../../prop-types/prop-types";

const playerScreenProps = {
  movie: movieProps,
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string
    })
  }),
  movies: PropTypes.arrayOf(movieProps)
};

export default playerScreenProps;
