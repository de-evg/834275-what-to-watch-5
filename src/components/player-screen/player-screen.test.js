import React from "react";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import renderer from "react-test-renderer";
import {PlayerScreen} from "./player-screen";
import {Router, Route} from "react-router-dom";
import {createBrowserHistory} from "history";

describe(`Render PlayerScreen`, () => {
  const mockStore = configureStore([]);
  let store = null;
  let PlayerScreenComponent = null;
  const history = createBrowserHistory();
  history.push(`/player/1`);

  const movie = {
    id: 0,
    title: `The Grand Budapest`,
    genre: `Horror`,
    release: 2014,
    posterURL: `/img/the-grand-budapest-hotel-poster.jpg`,
    previewURL: `/img/bg-the-grand-budapest-hotel.jpg`,
    rating: `8,9`,
    ratingLevel: `Good`,
    ratingCount: 10,
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege. Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,
    director: `Wes Andreson`,
    actors: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`],
    videoURL: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    isInWatchList: Math.random() <= 1,
    runTime: 146
  };

  beforeEach(() => {
    store = mockStore({
      DATA: {
        movies: [movie]
      }
    });

    store.dispatch = jest.fn();

    const withProps = (Component) => {
      return <Component
        movies={[movie]}
        match={{params: {id: `0`}}}
      />;
    };

    PlayerScreenComponent = renderer
      .create(
          <Provider store={store}>
            <Router history={history}>
              <Route
                exact
                component={() => withProps(PlayerScreen)}
                path={`/player/:id`}
                route={`/player/1`}
              />
            </Router>
          </Provider>
      );
  });

  it(`Should PlayerScreen connected to store render correctly`, () => {
    expect(PlayerScreenComponent.toJSON()).toMatchSnapshot();
  });
});
