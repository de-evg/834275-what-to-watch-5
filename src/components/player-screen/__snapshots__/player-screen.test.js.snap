// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render PlayerScreen Should PlayerScreen connected to store render correctly 1`] = `
<div
  className="player"
>
  <video
    autoPlay={true}
    className="player__video"
    onCanPlay={[Function]}
    onClick={[Function]}
    onTimeUpdate={[Function]}
    poster="/img/bg-the-grand-budapest-hotel.jpg"
    src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
  />
  <a
    className="player__exit"
    href="/"
    onClick={[Function]}
    style={
      Object {
        "textDecoration": "none",
      }
    }
    type="button"
  >
    Exit
  </a>
  <div
    className="player__controls"
  >
    <div
      className="player__controls-row"
    >
      <div
        className="player__time"
      >
        <progress
          className="player__progress"
          max="100"
          value={0}
        />
        <div
          className="player__toggler"
          style={
            Object {
              "left": "0%",
            }
          }
        >
          Toggler
        </div>
      </div>
      <div
        className="player__time-value"
      >
        0:2:26
      </div>
    </div>
    <div
      className="player__controls-row"
    >
      <button
        className="player__play"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="21"
          viewBox="0 0 14 21"
          width="14"
        >
          <use
            xlinkHref="#pause"
          />
        </svg>
        <span>
          Pause
        </span>
      </button>
      <div
        className="player__name"
      >
        The Grand Budapest
      </div>
      <button
        className="player__full-screen"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="27"
          viewBox="0 0 27 27"
          width="27"
        >
          <use
            xlinkHref="#full-screen"
          />
        </svg>
        <span>
          Full screen
        </span>
      </button>
    </div>
  </div>
</div>
`;
